errors/multi_test.go

```diff
@@ -1,121 +1,279 @@
-			err: Append(myErrState, Wrap(Append(myErrMsg, Wrap(ErrNotFound, "wrapped")),
-				"wrapped")),
-			exp: ErrNotFound,
+		"multi error is all of represented errors: not found": {
+			Kind:   ErrNotFound,
+			Tested: Append(ErrInput, ErrNotFound, ErrUnauthorized),
+			WantIs: true,
+		},
+		"multi error is all of represented errors: unauthorized": {
+			Kind:   ErrUnauthorized,
+			Tested: Append(ErrInput, ErrNotFound, ErrUnauthorized),
+			WantIs: true,
+		},
+		"multi error is all of represented errors but not others": {
+			Kind:   ErrPanic,
+			Tested: Append(ErrInput, ErrNotFound, ErrUnauthorized),
+			WantIs: false,
+		},
+		"multi error is all of represented errors when wrapped": {
+			Kind: ErrNotFound,
+			Tested: Append(
+				Wrap(ErrInput, "foo"),
+				Wrap(ErrNotFound, "bar"),
+			),
+			WantIs: true,
+		},
+		"multi error is all of represented errors when nested": {
+			Kind: ErrNotFound,
+			Tested: Append(
+				ErrInput,
+				Append(
+					ErrState,
+					Append(ErrSchema, ErrNotFound),
+				),
+			),
+
+			WantIs: true,
+		},
+		"multi error is only all of represented errors when nested": {
+			Kind: ErrEmpty,
+			Tested: Append(
+				ErrInput,
+				Append(
+					ErrState,
+					Append(ErrSchema, ErrNotFound),
+				),
+			),
+
+			WantIs: false,
+		},
+		"multi error is all of represented errors when nested and wrapped": {
+			Kind: ErrNotFound,
+			Tested: Wrap(Append(
+				ErrInput,
+				Wrap(Append(
+					ErrState,
+					Wrap(Append(ErrSchema, ErrNotFound), "inner"),
+				), "middle"),
+			), "most outer"),
+
+			WantIs: true,
 		},
 	}
 
-	for msg, spec := range specs {
-		t.Run(msg, func(t *testing.T) {
-			if !spec.exp.Is(spec.err) {
-				t.Errorf("expected %v to contain %v", spec.err, spec.exp)
+	for testName, tc := range cases {
+		t.Run(testName, func(t *testing.T) {
+			if is := tc.Kind.Is(tc.Tested); is != tc.WantIs {
+				t.Fatalf("unexpected result: %v", is)
 			}
 		})
 	}
 }
 
-func TestMultiErrIsEmpty(t *testing.T) {
-	specs := map[string]struct {
-		src multiErr
-		exp bool
+func TestMultiErrorMessageFormat(t *testing.T) {
+	cases := map[string]struct {
+		Err     error
+		WantMsg string
 	}{
-		"Single error": {src: multiErr{ErrNotFound}, exp: false},
-		"Empty":        {src: multiErr{}, exp: true},
-		"nil":          {src: nil, exp: true},
+		"nil multi error": {
+			// This is an unusual case and most likely a bug in the
+			// code if you do this, but let's test the result
+			// nevertheless.
+			Err:     (multiError)(nil),
+			WantMsg: "<nil>",
+		},
+		"empty multi error": {
+			// This is an unusual case and most likely a bug in the
+			// code if you do this, but let's test the result
+			// nevertheless.
+			Err:     multiError{},
+			WantMsg: "<nil>",
+		},
+		"single error": {
+			Err:     Append(errors.New("my msg")),
+			WantMsg: "my msg",
+		},
+		"two errors": {
+			Err: Append(
+				errors.New("first msg"),
+				errors.New("second msg"),
+			),
+			WantMsg: `2 errors occurred:
```

Comment:
This is a change compared to the original implementation. I have decided to not produce JSON for two reasons:
1. We use plain text everywhere. Introducing JSON assumes that there is a machine parsing and understanding those messages. Using JSON for multi-error case is also inconsistent with other error messages. As a result, some error messages were producing a plain text `Log` format and others some JSON structure.
2. Previous implementation was type casing and [destroying](https://github.com/iov-one/weave/blob/master/errors/multi.go#L43) some errors and implementation casting in order to maintain a flat list of errors. As explained [in another comment](https://github.com/iov-one/weave/pull/776#discussion_r296188862) this is not a good solution. Because we must use a tree structure, JSON serialized result might be very complex. Unreadable for humans, hard to deal with for machines.

Another big different is that the previous implementation was expsing the error code of each error. The new implementation does not. It coud include it in the message string. As always, for complex errors structure, this will produce a hard to undestand result anyway.